File: Scala/src/main/scala/AlgorithmicHeights/MS.scala
package AlgorithmicHeights

object MS {
  import scala.annotation.tailrec

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def mergeSortedLists(list1: List[Int], list2: List[Int]): List[Int] = {
    @tailrec
    def loop(acc: List[Int], xs: List[Int], ys: List[Int]): List[Int] = (xs, ys) match {
      case (Nil, Nil) => acc.reverse
      case (x :: xss, Nil) => loop(x :: acc, xss, ys)
      case (Nil, y :: yss) => loop(y :: acc, xs, yss)
      case (x :: xss, y :: yss) => if (x <= y) loop(x :: acc, xss, ys) else loop(y :: acc, xs, yss)
    }

    loop(Nil, list1, list2)
  }

  def mergeSort(array: List[Int], n: Int): List[Int] = {
    if (n <= 1) array
    else {
      val k: Int = n / 2
      val (first, second): (List[Int], List[Int]) = array.splitAt(k)
      val sortedFirst: List[Int] = mergeSort(first, k)
      val sortedSecond: List[Int] = mergeSort(second, n - k)
      mergeSortedLists(sortedFirst, sortedSecond)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: List[Int] = convertToIntList(reader.next())
    val result: List[Int] = mergeSort(array, n)
    println(result.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/ms.ml
open Batteries


let convert_to_intlist (line : string) : int list =
    List.map int_of_string Str.(line |> split (regexp " "))


let merge_sorted_lists (lst1 : int list) (lst2 : int list) : int list =
    let rec loop acc xs ys = match (xs, ys) with
        | ([], []) -> List.rev acc
        | (x :: xss, []) -> loop (x :: acc) xss ys
        | ([], y :: yss) -> loop (y :: acc) xs yss
        | (x :: xss, y :: yss) -> if x <= y then loop (x :: acc) xss ys else loop (y :: acc) xs yss
    in loop [] lst1 lst2


let rec merge_sort (xs : int list) (n : int) : int list =
    if n <= 1 then xs
    else
        let k = n / 2 in
        let xs1, xs2 = List.split_at k xs in
        let sorted1 = merge_sort xs1 k in
        let sorted2 = merge_sort xs2 (n - k) in
        merge_sorted_lists sorted1 sorted2


let () =
    let n = read_int () in
    let xs = convert_to_intlist (read_line ()) in
    let result = merge_sort xs n in
    result |> List.map string_of_int |> String.concat " " |> print_endline
